sunpy/physics/differential_rotation.py

```diff
@@ -84,138 +88,316 @@ def diff_rot(duration: u.s, latitude: u.deg, rot_type='howard', frame_time='side
+    over which to the amount of solar rotation is calculated. Note that using
+    the "time" keyword assumes that the new observer is on the Earth. This may
+    be a reasonable assumption depending on the application.
+
+    Either the "observer" or "time" argument must be specified, but both
+    cannot be specified at the same time and both cannot be None.
+
+    Parameters
+    ----------
+    initial_obstime : `~astropy.time.Time`
+        The initial time before solar rotation has been applied.
+
+    observer : `~astropy.coordinates.BaseCoordinateFrame`, `~astropy.coordinates.SkyCoord`, None
+        The location of the new observer in space and time (the observer must have an
+        interpretable obstime property).
+
+    time : `~astropy.time.Time`, `~astropy.time.TimeDelta`, `~astropy.units.Quantity`, None
+        Used to define the duration over which the amount of solar rotation is
+        calculated.  If 'time' is an `~astropy.time.Time` then the time interval is
+        "time - initial_obstime"; if 'time' is `~astropy.time.TimeDelta` or
+        `~astropy.units.Quantity` then the calculation is "initial_obstime + time".
+
+    Returns
+    -------
+    new_observer : `~astropy.coordinates.SkyCoord`, `~astropy.coordinates.BaseCoordinateFrame`
+        The position of the observer in space and time. If the "time" keyword is used
+        the output is an `~astropy.coordinates.SkyCoord`. If the "observer" keyword
+        is not None the output has the same type as the "observer" keyword.  In all cases
+        the output is specified in the heliographic Stonyhurst coordinate system.
+    """
+    # Check the input and create the new observer
+    if (observer is not None) and (time is not None):
+        raise ValueError("Either the 'observer' or the 'time' keyword must be specified, but not both simultaneously.")
+    elif observer is not None:
+        # Check that the new_observer is specified correctly.
+        if not (isinstance(observer, (BaseCoordinateFrame, SkyCoord))):
+            raise ValueError(
+                "The 'observer' must be an astropy.coordinates.BaseCoordinateFrame or an astropy.coordinates.SkyCoord.")
+        if observer.obstime is None:
+            raise ValueError("The observer 'obstime' property must not be None.")
+        new_observer = observer
+    elif time is not None:
+        warnings.warn("Using 'time' assumes an Earth-based observer.")
+        if isinstance(time, TimeDelta) or isinstance(time, u.Quantity):
+            new_observer_time = initial_obstime + time
+        else:
+            new_observer_time = parse_time(time)
+
+        new_observer = get_body("earth", new_observer_time)
+    else:
+        raise ValueError("Either the 'observer' or the 'time' keyword must not be None.")
+
+    return new_observer.transform_to(HeliographicStonyhurst)
+
+
+def solar_rotate_coordinate(coordinate, observer=None, time=None, **diff_rot_kwargs):
```

Comment:
Just like we were discussing today, I don't like that the time can technically be passed in two different ways.

Reply:
Also it seems like it is not necessary to have a function which rotates both the Sun and the observer at the same time. Is there only a differential rotate function without the ability to rotate the observer? Seems like we should have that as well.

Reply:
Yes, make the new observer the same as map observer, except specify the time.  That way the observer does not move.  I was going to put that in the docs.  I can't think of a use case for this though, but that doesn't mean anything.

I originally wanted this function to take observers only, and the observer specification would carry the new observer time.  I was persuaded that allowing the user to specify either an observer or a time would be more user friendly.  This explains the need for the (hidden) function _get_new_observer.  This non-user facing function is used to interpret the observer and time keywords passed into `solar_rotate_coordinate`.